### ISSUE TITLE
Enable inference of input types when using FnToProduct.

### ISSUE THREAD

**Issue by AlecZorab** [milessabin/shapeless#350]
From the shapeless 2.0 feature overview:

```scala
import syntax.std.function._
import ops.function._

def applyProduct[P <: Product, F, L <: HList, R](p: P)(f: F)
  (implicit gen: Generic.Aux[P, L], fp: FnToProduct.Aux[F, L => R]) =
    f.toProduct(gen.to(p))

scala> applyProduct(1, 2)((_: Int)+(_: Int))
res0: Int = 3

scala> applyProduct(1, 2, 3)((_: Int)*(_: Int)*(_: Int))
res1: Int = 6
```
it would be nice to be able to provide `(_ + _)' and have the parameters inferred.

Similarly, given an `F[A <: Hlist]` and a suitable functor instance for `F`, it would be nice to be able to call `map` with a lambda of the appropriate arity and have the types inferred correctly.

**Comment by milessabin** [comment]
It looks like all this needs is a contravariant `FnToProduct`. If you like to experiment with making the primary definition in shapeless contravariant (it's not quite a simple as just adding a `-`) I'd be happy to include that in 2.3.0,

```scala
import shapeless._

import ops.function._

// Contravariant version of shapeless's FnToProduct
trait ContraFnToProduct[-F] {
  type Out
  def apply(f: F): Out
}

object ContraFnToProduct {
  type Aux[F, Out0] = ContraFnToProduct[F] { type Out = Out0 }

  implicit def fnp2[A, B, C, Res]: Aux[(A, B, C) => Res, (A :: B :: C :: HNil) => Res] =
    new ContraFnToProduct[(A, B, C) => Res] {
      type Out = (A :: B :: C :: HNil) => Res
      def apply(f: (A, B, C) => Res): Out = (l: A :: B :: C :: HNil) => f(l.head, l.tail.head, l.tail.tail.head)
    }
}

object Test extends App {
  class HListSyntax[A <: HList, F <: AnyRef](a: A) {
    def $$[U](f: F)(implicit cftp: ContraFnToProduct.Aux[f.type, A => U]): U = cftp(f)(a)
  }
  implicit def mkSyntax[A <: HList, F <: AnyRef](a: A)
    (implicit ffp: FnFromProduct.Aux[A => Any, F]): HListSyntax[A, F] =
      new HListSyntax[A, F](a)

  val res = (2 :: "a" :: 1.3 :: HNil) $$ ((i, s, d) => (s * i, d * i)) // Function argument types inferred
  assert((res: (String, Double)) == ("aa", 2.6))
}
```

**Comment by AndreasKostler** [comment]
I have a patch for this - however, I don't quite understand how making `F` contravariant aids with type inference in this case.

**Comment by milessabin** [comment]
I have to confess that I'm not entirely sure either. As a general rule indexing a type class by a contravariant type parameter allows it to be used in context where an over-precise type has been inferred ... I'm not quite sure how that plays out in this case.

**Comment by AndreasKostler** [comment]
:) Glad it's not just me ;)

**Comment by AndreasKostler** [comment]
The change is trivial btw, I've just got another pr open atm.

**Comment by milessabin** [comment]
Fixed in 48d8be07a62815d9e6dddb6aa370f95d26353fca and 048bc8753e1726ea6497b6553621cf276efa7c70.
